### PR TITLE
Recommendation algorithm finetuning

### PR DESCRIPTION
**Pull request by hernandezvictoria**
## Description

- added checks for heavily liked/disliked products -- will increase/decrease score for repetitive brands or ingredients within liked/disliked products
- updated weights for products with no ingredients

## Milestones
- TC 1

## Resources
- asked metamate for ideas on edge cases, generally unhelpful

## Test Plan
- added a few more test cases in test.js and compared scores before and after making changes to the algorithm
- for example, got score for product with no ingredients for 3 different users; then tweaked the algorithm so that products with no ingredients have different weights; verified that score went up for all 3 users
- followed similar procedure with the updated liked/disliked system
- also individually tested helper methods (parseDislikedProducts and parseLikedProducts) to verify that they return the right brands and ingredients who had high frequencies.

### CODE REVIEW

File: backend/routes/scoring-helper-functions.js [hunk 1]
```diff
@@ -1,6 +1,54 @@
 var jaccard = require('jaccard');
 const { SkinTypes, SkinConcerns, ProductTypes } = require('../enums.js')
 
+const parseDislikedProducts = (dislikedProducts) => {
```
Comment (ryanwproulx) [inline backend/routes/scoring-helper-functions.js:4]: I would split this into two functions to make it a bit more clear

File: backend/routes/scoring-helper-functions.js [hunk 2]
```diff
@@ -1,6 +1,54 @@
...
+        ingredientFrequencies[ingredient] = (ingredientFrequencies[ingredient] || 0) + 1;
+        if(ingredientFrequencies[ingredient] >= 5) {
+            dislikedIngredients.add(ingredient);
+        }
+    }
+    // list of brand names that appeared greater than 3 times, list of ingredient ids that appeared greater than 5 times
+    return {dislikedBrands, dislikedIngredients};
+}
+
+const parseLikedProducts = (likedProducts) => {
```
Comment (ryanwproulx) [inline backend/routes/scoring-helper-functions.js:28]: Could split this into 2 functions too

Comment (ryanwproulx) [inline backend/routes/scoring-helper-functions.js:28]: You could actually combine this with the first function to be something like:

const parseLikesDislikes = (products) => {
    const productsBrands = products.map(product => product.brand);
    const brandsSet = new Set();
    const brandFrequencies = {}; // count of each brand
    for (const brand of likedDislikesBrands) {
        brandFrequencies[brand] = (brandFrequencies[brand] || 0) + 1;
        if(brandFrequencies[brand] >= 3) {
            brandsSet.add(brand);
        }
    }
    
    return brandsSet
    
    
    
   And then can do the same for the ingredients

File: backend/routes/scoring-helper-functions.js (hunk 1 above)
Comment (ryanwproulx) [reply backend/routes/scoring-helper-functions.js:4]: see below comment on combining these functions

File: backend/routes/scoring-helper-functions.js [hunk 3]
```diff
@@ -1,6 +1,54 @@
...
+            likedBrands.add(brand);
+        }
+    }
+
+    const likedProductsIngredients = likedProducts.map(product => product.ingredients.map(ingredient => ingredient.id)).flat();
+    const likedIngredients = new Set();
+    const ingredientFrequencies = {}; // count of each brand
+    for (const ingredient of likedProductsIngredients) {
+        ingredientFrequencies[ingredient] = (ingredientFrequencies[ingredient] || 0) + 1;
+        if(ingredientFrequencies[ingredient] >= 4) {
```
Comment (ryanwproulx) [inline backend/routes/scoring-helper-functions.js:44]: why is # disliked ingredients different to liked?

File: backend/routes/scoring-helper-functions.js [hunk 4]
```diff
@@ -103,15 +162,30 @@ const computeProductScore = (product, lovedProducts, dislikedProducts, userSkinT
         dislikedProductOverlapScore += (dislikedProductIngredientSimilarityScore / dislikedProducts.length); // average jaccard score of disliked products
     }
 
+    for(const ingredientId of dislikedIngredients) {
+        if(product.ingredients.some(i => i.id === ingredientId)) {
```
Comment (ryanwproulx) [inline backend/routes/scoring-helper-functions.js:166]: Correct me if I am wrong, but dislikedIngredients is a set, so would it be faster to loop over product.ingredients and do an O(1) check if it is in dislikedIngredients?

File: backend/routes/scoring-helper-functions.js [hunk 5]
```diff
@@ -103,15 +162,30 @@ const computeProductScore = (product, lovedProducts, dislikedProducts, userSkinT
...
+            dislikedProductOverlapScore -= 0.5; // penalize extra for disliked ingredient
+        }
+    }
+    lovedProductOverlapScore = Math.min(lovedProductOverlapScore, 2); // cap at 2
+    dislikedProductOverlapScore = Math.max(dislikedProductOverlapScore, -2); // cap at -2
     const bonusScore = lovedProductOverlapScore + dislikedProductOverlapScore;
 
     // ========== combine all scores ===========
     let weights = {}; // can adjust weights further
```
Comment (ryanwproulx) [inline backend/routes/scoring-helper-functions.js:175]: love that they are all in one place!

**Review by ryanwproulx — APPROVED** [review]

File: backend/routes/scoring-helper-functions.js (hunk 4 above)
Comment (hernandezvictoria) [reply backend/routes/scoring-helper-functions.js:166]: so true! that would be a good use of the set data structure

File: backend/routes/scoring-helper-functions.js (hunk 2 above)
Comment (hernandezvictoria) [reply backend/routes/scoring-helper-functions.js:28]: this is a really good soln! thanks ryan!
